### PR TITLE
Support Script Context Stateful Factory in Painless

### PR DESCRIPTION
**Pull request by jdconrad**
As the title says.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java [hunk 1]
```diff
@@ -158,20 +159,126 @@ public Loader run() {
...
+        String interfaceBase = Type.getType(context.statefulFactoryClazz).getInternalName();
+        String className = interfaceBase + "$StatefulFactory";
+        String classInterfaces[] = new String[] { interfaceBase };
+
+        ClassWriter writer = new ClassWriter(classFrames);
+        writer.visit(WriterConstants.CLASS_VERSION, classAccess, className, null, OBJECT_TYPE.getInternalName(), classInterfaces);
+
+        Method newFactory = null;
+
+        for (Method method : context.factoryClazz.getMethods()) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: As a followup I think we should have the ScriptContext validate this method exists, so then we won't need to look it up here, but just grab the Method from the context.

Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: And the same for the newInstance method on the stateful factory.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java [hunk 2]
```diff
@@ -158,20 +159,126 @@ public Loader run() {
...
+            }
+        }
+
+        org.objectweb.asm.commons.Method instance = new org.objectweb.asm.commons.Method(newInstance.getName(),
+            MethodType.methodType(newInstance.getReturnType(), newInstance.getParameterTypes()).toMethodDescriptorString());
+
+        List<Class<?>> parameters = new ArrayList<>(Arrays.asList(newFactory.getParameterTypes()));
+        parameters.addAll(Arrays.asList(newInstance.getParameterTypes()));
+
+        org.objectweb.asm.commons.Method constru = new org.objectweb.asm.commons.Method("<init>",
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: As part of that same followup, we should also validate a ctor exists on the script class with the appropriate arguments from the factory methods.

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/FactoryTests.java [hunk 3]
```diff
@@ -30,13 +30,56 @@
...
+            StatefulFactory newFactory(int x, int y);
+        }
+
+        public static final ScriptContext<StatefulFactoryTestScript.Factory> CONTEXT =
+            new ScriptContext<>("test", StatefulFactoryTestScript.Factory.class);
+    }
+
+    public void testStatefulFactory() {
+        StatefulFactoryTestScript.Factory factory = scriptEngine.compile(
+            "stateful_factory_test", "test + x + y", StatefulFactoryTestScript.CONTEXT, Collections.emptyMap());
```
Comment (rjernst) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/FactoryTests.java]: What is "test" in this script?

**Review by rjernst — APPROVED** [review]
Looks good, I left a few suggestions for followup.

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/FactoryTests.java (hunk 3 above)
Comment (rjernst) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/FactoryTests.java]: Nevermind! I see it now.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java (hunk 1 above)
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: Agreed.  As much as possible should be on the context.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java (hunk 2 above)
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: Agreed.

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/FactoryTests.java (hunk 3 above)
Comment (jdconrad) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/FactoryTests.java]: test is the parameter passed into the script method execute.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!  Will merge once the CI passes.

**Pull request merged**